### PR TITLE
Re-enable "Flush SP cache after SP change"

### PR DESCRIPTION
**Pull request by jamesagnew**
Reverts hapifhir/hapi-fhir#4614

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/JpaResourceDaoSearchParameter.java [hunk 1]
```diff
@@ -42,6 +44,23 @@
...
+
+		/*
+		 * After we commit, flush the search parameter cache. This only helps on the
+		 * local server (ie in a cluster the other servers won't be flushed) but
+		 * the cache is short anyhow, and  flushing locally is better than nothing.
+		 * Many use cases where you would create a search parameter and immediately
+		 * try to use it tend to be on single-server setups anyhow, e.g. unit tests
+		 */
+		if (!shouldSkipReindex(theRequestDetails)) {
+			TransactionSynchronizationManager.registerSynchronization(new TransactionSynchronization() {
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/JpaResourceDaoSearchParameter.java:56]: question: Can this afterCommit hook get registered several times, and force refresh several times?  Do we need to debounce it?

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchCustomSearchParamTest.java [hunk 2]
```diff
@@ -1752,14 +1753,15 @@ public void testCompositeWithInvalidTarget() {
...
 			.setDefinition("http://foo");
-		mySearchParameterDao.create(sp);
 
 		IAnonymousInterceptor interceptor = mock(IAnonymousInterceptor.class);
 		myInterceptorRegistry.registerAnonymousInterceptor(Pointcut.JPA_PERFTRACE_WARNING, interceptor);
 
 		try {
-			mySearchParamRegistry.forceRefresh();
+			// Test
+			mySearchParameterDao.create(sp);
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchCustomSearchParamTest.java:1762]: hooray for easier testing.

**Review by michaelabuckley — APPROVED** [review]
Approved with one question about multiple calls forcing refresh multiple times.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/TaskChunker.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * This utility takes an input collection, breaks it up into chunks of a
+ * given maximum chunk size, and then passes those chunks to a consumer for
+ * processing. Use this to break up large tasks into smaller tasks.
+ *
+ * @since 6.6.0
+ * @param <T> The type for the chunks
+ */
+public class TaskChunker<T> {
+
+	public void chunk(Collection<T> theInput, int theChunkSize, Consumer<List<T>> theBatchConsumer) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/TaskChunker.java:38]: Those are close but I'd say not the same (I quite like the fact that one of the inputs to QueryChunker was a consumer for the chunks, which is why i wanted to refactor that into something generic)

Have added tests

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/JpaResourceDaoSearchParameter.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/JpaResourceDaoSearchParameter.java:56]: Good call - i'll add a guard against that

**Comment by iyt-trifork** [comment]
Im currently having issues with this commit. 
In my HAPI-FHIR setup, i perform a package install outside a transaction, and during a transaction synchronization at ca.uhn.fhir.jpa.dao.JpaResourceDaoSearchParameter, i get the following exception:

`
ca.uhn.fhir.rest.server.exceptions.InternalErrorException: HAPI-2223: No existing transaction found for transaction marked with propagation 'mandatory'
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.invokeCallableAndHandleAnyException(HapiTransactionService.java:462)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService$ExecutionBuilder.lambda$execute$1(HapiTransactionService.java:411)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.executeInExistingTransaction(HapiTransactionService.java:448)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.doExecute(HapiTransactionService.java:209)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService$ExecutionBuilder.execute(HapiTransactionService.java:419)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService$ExecutionBuilder.execute(HapiTransactionService.java:412)
	at ca.uhn.fhir.jpa.search.SynchronousSearchSvcImpl.executeQuery(SynchronousSearchSvcImpl.java:107)
	at ca.uhn.fhir.jpa.search.SearchCoordinatorSvcImpl.registerSearch(SearchCoordinatorSvcImpl.java:361)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.search(BaseHapiFhirResourceDao.java:1591)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.search(BaseHapiFhirResourceDao.java:1567)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.search(BaseHapiFhirResourceDao.java:1561)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao$$FastClassBySpringCGLIB$$5578fb67.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:137)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at ca.uhn.fhir.jpa.dao.JpaResourceDaoSearchParameter$$EnhancerBySpringCGLIB$$5546421f.search(<generated>)
	at ca.uhn.fhir.jpa.dao.DaoSearchParamProvider.search(DaoSearchParamProvider.java:39)
	at ca.uhn.fhir.jpa.searchparam.registry.SearchParamRegistryImpl.rebuildActiveSearchParams(SearchParamRegistryImpl.java:165)
	at ca.uhn.fhir.jpa.searchparam.registry.SearchParamRegistryImpl.handleChange(SearchParamRegistryImpl.java:339)
	at ca.uhn.fhir.jpa.cache.ResourceChangeListenerCacheRefresherImpl.compareLastVersionMapToNewVersionMapAndNotifyListenerOfChanges(ResourceChangeListenerCacheRefresherImpl.java:198)
	at ca.uhn.fhir.jpa.cache.ResourceChangeListenerCacheRefresherImpl.notifyListener(ResourceChangeListenerCacheRefresherImpl.java:162)
	at ca.uhn.fhir.jpa.cache.ResourceChangeListenerCacheRefresherImpl.refreshCacheAndNotifyListener(ResourceChangeListenerCacheRefresherImpl.java:145)
	at ca.uhn.fhir.jpa.cache.ResourceChangeListenerCache.lambda$refreshCacheAndNotifyListenersWithRetry$0(ResourceChangeListenerCache.java:145)
	at ca.uhn.fhir.jpa.searchparam.retry.Retrier.lambda$runWithRetry$0(Retrier.java:87)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209)
	at ca.uhn.fhir.jpa.searchparam.retry.Retrier.runWithRetry(Retrier.java:87)
	at ca.uhn.fhir.jpa.cache.ResourceChangeListenerCache.refreshCacheAndNotifyListenersWithRetry(ResourceChangeListenerCache.java:148)
	at ca.uhn.fhir.jpa.cache.ResourceChangeListenerCache.refreshCacheWithRetry(ResourceChangeListenerCache.java:130)
	at ca.uhn.fhir.jpa.cache.ResourceChangeListenerCache.forceRefresh(ResourceChangeListenerCache.java:85)
	at ca.uhn.fhir.jpa.searchparam.registry.SearchParamRegistryImpl.forceRefresh(SearchParamRegistryImpl.java:271)
	at ca.uhn.fhir.jpa.dao.JpaResourceDaoSearchParameter$1.afterCommit(JpaResourceDaoSearchParameter.java:64)
	at org.springframework.transaction.support.TransactionSynchronizationUtils.invokeAfterCommit(TransactionSynchronizationUtils.java:135)
	at org.springframework.transaction.support.TransactionSynchronizationUtils.triggerAfterCommit(TransactionSynchronizationUtils.java:123)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.triggerAfterCommit(AbstractPlatformTransactionManager.java:936)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:782)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:711)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:152)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.doExecuteCallback(HapiTransactionService.java:335)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.doExecuteInTransaction(HapiTransactionService.java:230)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.doExecute(HapiTransactionService.java:215)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService$ExecutionBuilder.execute(HapiTransactionService.java:419)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService$ExecutionBuilder.execute(HapiTransactionService.java:412)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.update(BaseHapiFhirResourceDao.java:1799)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.update(BaseHapiFhirResourceDao.java:1761)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.update(BaseHapiFhirResourceDao.java:1756)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.update(BaseHapiFhirResourceDao.java:1743)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao$$FastClassBySpringCGLIB$$5578fb67.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:137)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at ca.uhn.fhir.jpa.dao.JpaResourceDaoSearchParameter$$EnhancerBySpringCGLIB$$5546421f.update(<generated>)
	at ca.uhn.fhir.jpa.packages.PackageInstallerSvcImpl.updateResource(PackageInstallerSvcImpl.java:375)
	at ca.uhn.fhir.jpa.packages.PackageInstallerSvcImpl.create(PackageInstallerSvcImpl.java:324)
	at ca.uhn.fhir.jpa.packages.PackageInstallerSvcImpl.install(PackageInstallerSvcImpl.java:223)
	at ca.uhn.fhir.jpa.packages.PackageInstallerSvcImpl.install(PackageInstallerSvcImpl.java:175)
	at com.trifork.ehealth.util.ResourceLoader.installPackage(ResourceLoader.java:159)
	at com.trifork.ehealth.util.ResourceLoader.installEhealthCoreImplementationGuide(ResourceLoader.java:136)
	at com.trifork.ehealth.util.ResourceLoader.populateResources(ResourceLoader.java:109)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at 
*snip*
Caused by: org.springframework.transaction.IllegalTransactionStateException: No existing transaction found for transaction marked with propagation 'mandatory'
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:362)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:595)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:382)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:137)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:145)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:220)
	at com.sun.proxy.$Proxy308.findByResourceIds(Unknown Source)
	at ca.uhn.fhir.jpa.search.builder.SearchBuilder.doLoadPids(SearchBuilder.java:943)
	at ca.uhn.fhir.jpa.search.builder.SearchBuilder.lambda$loadResourcesByPid$9(SearchBuilder.java:1071)
	at ca.uhn.fhir.util.TaskChunker.chunk(TaskChunker.java:49)
	at ca.uhn.fhir.jpa.util.QueryChunker.chunk(QueryChunker.java:39)
	at ca.uhn.fhir.jpa.search.builder.SearchBuilder.loadResourcesByPid(SearchBuilder.java:1071)
	at ca.uhn.fhir.jpa.search.SynchronousSearchSvcImpl.lambda$executeQuery$1(SynchronousSearchSvcImpl.java:192)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.invokeCallableAndHandleAnyException(HapiTransactionService.java:458)
	... 167 more


`

**Pull request merged**